pkg/querier/tripperware/metadata/get_series.go

```diff
@@ -0,0 +1,288 @@
+package metadata
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/querier/tripperware"
+	"github.com/cortexproject/cortex/pkg/querier/tripperware/queryrange"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/spanlogger"
+	"github.com/gogo/status"
+	jsoniter "github.com/json-iterator/go"
+	"github.com/opentracing/opentracing-go"
+	otlog "github.com/opentracing/opentracing-go/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/timestamp"
+	"github.com/weaveworks/common/httpgrpc"
+)
+
+var (
+	GetSeriesCodec tripperware.Codec = newGetSeriesCodec()
+	json                             = jsoniter.Config{
+		EscapeHTML:             false, // No HTML in our responses.
+		SortMapKeys:            true,
+		ValidateJsonRawMessage: true,
+	}.Froze()
+)
+
+type PrometheusGetSeriesResponseExtended struct {
```

Comment:
Can we use `repeated cortexpb.LabelPair` for this?

Reply:
I don't think this is possible. If I do repeated LabelPair, I'll get something like this:

```
type PrometheusGetSeriesResponse struct {
	Status    string                                  `protobuf:"bytes,1,opt,name=Status,proto3" json:"status"`
	ErrorType string                                  `protobuf:"bytes,3,opt,name=ErrorType,proto3" json:"errorType,omitempty"`
	Error     string                                  `protobuf:"bytes,4,opt,name=Error,proto3" json:"error,omitempty"`
	Data      []*cortexpb.LabelPair                   `protobuf:"bytes,5,rep,name=Data,proto3" json:"data,omitempty"`
	Headers   []*tripperware.PrometheusResponseHeader `protobuf:"bytes,6,rep,name=Headers,proto3" json:"-"`
}
```

```
type LabelPair struct {
	Name  []byte `protobuf:"bytes,1,opt,name=name,proto3" json:"name,omitempty"`
	Value []byte `protobuf:"bytes,2,opt,name=value,proto3" json:"value,omitempty"`
}
```

What we really want is something like `repeated repeated cortexpb.LabelPair` which is not allowed.

Reply:
We cannot add a new message type for repeated label pairs in the protobuf and then use repeated for the new type?

Reply:
The LabelPair also won't work because it has `Name` and `Value` fields which we don't want in the JSON output.

However, if we do similar nested message using a map:

```
message PrometheusGetSeriesResponse {
  string Status = 1 [(gogoproto.jsontag) = "status"];
  string ErrorType = 2 [(gogoproto.jsontag) = "errorType,omitempty"];
  string Error = 3 [(gogoproto.jsontag) = "error,omitempty"];
  repeated tripperware.PrometheusResponseHeader Headers = 4 [(gogoproto.jsontag) = "-"];
  repeated Metric Data = 5 [(gogoproto.jsontag) = "data"]
}

message Metric {
  map[string]string labels = 1 [(gogoproto.jsontag) = "labels"];
}
```

This will be serialized to:
```
{
  "status": "Success",
  "data": [
      labels: {"__name__": "metric1", "label": "value"}
  ]
}
```

Instead what we want is:
```
{
  "status": "Success",
  "data": [
     "__name__": "metric1", "label": "value"}
  ]
}
```

Reply:
Yeah we definitely cannot use that directly but it should work with a custom marshal/unmarshal implementation for the series response.
Anyway this is just one option. If we have other workable options then it is also ok.